SearchParser.js

```diff
@@ -0,0 +1,49 @@
+/**
+ * SearchParser
+ * @param {Array} patterns  Enumerates params and the patterns that match them
+ * @param {boolean} options.returnAllMatches  If true return all matches (not just first)
+ */
+
+'use strict';
+
+module.exports = function SearchParser(patterns, options) {
+    this.patterns = patterns;
+    this.returnAllMatches = options ? options.returnAllMatches : false;
+    /**
+     * Run regex match against supplied search input
+     * @param  {string} input
+     * @return {Object}       Match result, with supplied params as keys
+     *                        and matched strings as values
+     */
+    this.match = function (input) {
+        var matches = {};
+
+        if (input.length) {
+            this.patterns.some(function (obj) {
```

Comment:
could `obj` be renamed to something more descriptive? `pattern`, or `schemaPart`, for example.